README.md
# Kyoko Bot for Discord
![](https://img.shields.io/badge/python-3.9-blue) ![](https://img.shields.io/badge/discord.py-2.0.1-7289da)

Simple Anime Bot for Discord to search for and watch anime.

## Commands
- `!anime <name>`: searches MyAnimeList for a certain anime and returns top 9 results. Select a result to learn more about the anime.
- `!manga <name>` : searches MyAnimeList for a certain manga nad returns top 9 results. Select a result to learn more about the manga. 
- `!profile <profile name>`: searches for a MyAnimeList user and returns their stats



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [1.2.0] 2023-07-18
### Added
- Added `!manga` command  

## [1.1.1] 2022-12-28
### Added
- `!profile` command
- Returns anime stats properly, still need to add manga stats

### Changed
- Created separate `AnimeSearch.py` and `ProfileSearch.py` files for their respective functions to clean up `main` a little bit

## [1.1.0] 2022-12-26
### Added
- Anime search now also shows image

## [1.0.9] 2022-12-26
### Changed
- Fixed off-by-one search error on line 70
- Made docstrings more readable

### Added
- New selection edits message instead of sending new one
- Reacting to message now unreacts immediately (for aesthetics)

## [1.0.8] - 2022-10-23
### Changed
- Fixed the description thing - reacting to the message now shows the corresponding description

## [1.0.7] - 2022-10-22
### Added
- Displays description upon reaction (granted not the right one, but hey it's something)
- `get_url()` function
- `get_description` function

## [1.0.6] - 2022-10-10
### Changed
- Bug displayed results starting from the second result instead of top result
- Needed to make some variables global, working on a way around this

## [1.0.5] - 2022-10-10
### Changed
- `animeSearch` is now it's own function to make code a bit more modular

## [1.0.4] - 2022-10-09
### Added
- Numbering for results
- Reactions

### Changed
- Returns top 5 results instead of 10

### Removed
- `sentenceCase()` function (useless)

## [1.0.3] - 2022-10-09
### Changed
- `!anime` command returns basic HTML parsing, instead of just URL


## [1.0.2] - 2022-10-08
### Changed
- Uses embed instead of message

### Added
- `SentenceCase()` method

## [1.0.1] - 2022-10-08
### Changed
- `!anime` searches specifically for anime instead of all mediums
- Added docstrings for functions
- If condition uses channel variable

### Added
- `CHANGELOG.md`

## [1.0.0] - 2022-10-08
### Added
- Initial release

test.py
from ast import keyword
import discord
from discord import SyncWebhook
from discord.utils import get
from dotenv import load_dotenv
import os
import requests
from bs4 import BeautifulSoup

def get_description(url):
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    description_html = soup.find_all('p', attrs={'itemprop':'description'})

    for block in description_html:
        print(block)

    description = str(description_html)[27:-5].replace('<br>', '\n').replace('<br/>', '')

    # print(description)
    return description

def get_url(keyword):
    ''' Takes in a keyword and returns the url from that keyword '''
    r = requests.get(f'https://myanimelist.net/anime.php?q={keyword}&cat=anime')
    soup = BeautifulSoup(r.text, 'html.parser')
    links = soup.find_all('a')

    for link in links:
        print(link.get('href'))
    # print(links[93].get('href'))
    # return links[93].get('href')

get_url('one%20piece')
# get_description('https://myanimelist.net/anime/31490/One_Piece_Film__Gold')

main.py
import discord
from discord import SyncWebhook
from discord.utils import get
from dotenv import load_dotenv
import os
import requests
from bs4 import BeautifulSoup
from urlextract import URLExtract

from ProfileSearch import *
from AnimeSearch import *
from MangaSearch import *

# loading in token and webhook from .env
load_dotenv()
TOKEN = os.getenv("TOKEN")
webhook = os.getenv("WEBHOOK")

# global variables
client = discord.Client(intents=discord.Intents.all())
results = []
keyword = ""
first_reaction = True
isMangaSearch = False
isAnimeSearch = False


@client.event
async def on_ready():
    """Sends up message to console"""
    print('Bot started with username: {0.user}'.format(client))

@client.event
async def on_message(message):
    """Listens for message event and reads command

    Args:
        message (message): the latest message sent
    """
    global first_reaction
    global isMangaSearch
    global isAnimeSearch

    user_message = str(message.content) # gets content of message

    # only read human messages
    if message.author == client.user:
        return

    # !anime command
    if user_message.lower().startswith('!anime'):
        # every new command, send a new embed (see on_reaction_add method)
        first_reaction = True
        isAnimeSearch = True
        await animeSearch(user_message, message)

    if user_message.lower().startswith('!profile'):
        await profileSearch(user_message, message)

    if user_message.lower().startswith('!manga'):
        first_reaction = True
        isMangaSearch = True
        await mangaSearch(user_message, message)

    return

async def animeSearch(user_message, message):
    """Takes in user message as input and searches MyAnimeList for those keywords in animeSearch

    Args:
        user_message (string): CONTENT of the message
        message (message): message OBJECT
    """
    global results
    global keyword
    keyword = user_message[7:].replace(" ", "%20") # parse the command out of the message, replace spaces with '%20'

    # parse MyAnimeList and return top 10 results
    r = requests.get(f'https://myanimelist.net/anime.php?q={keyword}&cat=anime')
    soup = BeautifulSoup(r.text, 'html.parser')
    results = soup.find_all('a', attrs={'class':'hoverinfo_trigger fw-b fl-l'})
    return_message = ""

    # take the top 9 results from MyAnimeList
    for i in range(0, 9, 1):
        results[i] = str(results[i].find('strong'))[8:-9]
        return_message += f'{i + 1}. {str(results[i])}\n'

    # send message as embed and set reaction buttons 1-9
    embed = discord.Embed(title=f'Results for: {keyword.replace("%20", " ")}', description=f"{return_message}", color=0x36509D)
    msg = await message.channel.send(embed=embed)
    reactions = ["1️⃣", "2️⃣", "3️⃣", "4️⃣", "5️⃣", "6️⃣", "7️⃣", "8️⃣", "9️⃣"]

    # add reactions
    for reaction in reactions:
        await msg.add_reaction(reaction)

async def mangaSearch(user_message, message):
    """Takes in user message as input and searches MyAnimeList for those keywords in mangaSearch

    Args:
        user_message (string): CONTENT of the message
        message (message): message OBJECT
    """
    global results
    global keyword
    keyword = user_message[7:].replace(" ", "%20") # parse the command out of the message, replace spaces with '%20'

    # parse MyAnimeList and return top 10 results
    r = requests.get(f'https://myanimelist.net/manga.php?q={keyword}&cat=manga')
    soup = BeautifulSoup(r.text, 'html.parser')
    results = soup.find_all('a', attrs={'class':'hoverinfo_trigger fw-b'})
    return_message = ""

    # take the top 9 results from MyAnimeList
    for i in range(0, 9, 1):
        results[i] = str(results[i].find('strong'))[8:-9]
        return_message += f'{i + 1}. {str(results[i])}\n'

    # send message as embed and set reaction buttons 1-9
    embed = discord.Embed(title=f'Results for: {keyword.replace("%20", " ")}', description=f"{return_message}", color=0x36509D)
    msg = await message.channel.send(embed=embed)
    reactions = ["1️⃣", "2️⃣", "3️⃣", "4️⃣", "5️⃣", "6️⃣", "7️⃣", "8️⃣", "9️⃣"]

    # add reactions
    for reaction in reactions:
        await msg.add_reaction(reaction)


async def profileSearch(user_message, message):
    """Searches for an MAL profile and returns its stats

    Args:
        user_message (string): the message from the user command containing the profile to search for
        message (string): message object
    """
    # parse the "!profile " part out
    profile_name = user_message[9:]
    profile_url = f'https://myanimelist.net/profile/{profile_name}'
    r = requests.get(profile_url)
    soup = BeautifulSoup(r.text, 'html.parser')

    # if the request returns a error 404 (profile not found)
    if r.ok:
        embed = discord.Embed(title="MyAnimeList Profile", url=profile_url, color=0x00DBFF)
        embed.set_author(name=profile_name)
        embed.set_thumbnail(url=getProfilePicture(profile_url))
        embed.add_field(name="Anime Stats:", value=f"Days watched: {getDays(profile_url, 0)}\nEpisodes: {getEpisodes(profile_url)}\nWatching: {getWatching(profile_url)}\n Completed: {getCompleted(profile_url)}", inline=True)
        embed.add_field(name="Manga Stats:", value=f"Days read: {getDays(profile_url, 1)}\nChapters: \nReading: \n Completed: ", inline=True)
    else:
        embed = discord.Embed(title="Profile not found. ", description="Please enter a valid profile name.", color=0xFF0000)
        embed.set_thumbnail(url='https://64.media.tumblr.com/1e8b63b8e33978d7d5ef5019f32c5930/aebf5a6e89fb27e4-db/s400x600/37381ac3b8ae4be3df56bf18d20fcebf0fc00aa0.png')

    await message.channel.send(embed=embed)

@client.event
async def on_reaction_add(reaction, user):
    """Expands on description when reaction is pressed

    Args:
        reaction (reaction): reaction object
        user (user): user who reacted
    """
    global keyword
    global results
    global first_reaction
    global send
    global isMangaSearch
    global isAnimeSearch

    # if the bot is the one reacting, do nothing
    if reaction.message.author == user:
        return
    else:
        emoji_mapping = {'1️⃣': 0, '2️⃣': 1, '3️⃣': 2, '4️⃣': 3, '5️⃣': 4, '6️⃣':5, '7️⃣':6, '8️⃣':7, '9️⃣':8}
        # send the selected name and corresponding description of the selected search result
        if reaction.emoji in emoji_mapping:
            #checks to see if the search is for a Anime or Manga
            await reaction.remove(user)
            index = emoji_mapping[reaction.emoji]

            if isMangaSearch:
                url = getMangaUrl(keyword, index)
                description = getMangaDescription(url)
                isMangaSearch = False
            elif isAnimeSearch:
                url = getUrl(keyword, index)
                description = getDescription(url)
                isAnimeSearch = False

            embed = discord.Embed(title=f'{results[index]}', url=url, description=description, color=0x36509D)
            embed.set_thumbnail(url=getImage(url))

            if first_reaction:
                send = await reaction.message.channel.send(embed=embed)
                first_reaction = False
            else:
                await send.edit(embed=embed)

    return

if __name__ == "__main__":
    try:
        client.run(TOKEN)
    except:
        print("\nERROR: Client can't be run with token. Is .env file properly imported?")

MangaSearch.py
import discord
from discord import SyncWebhook
from discord.utils import get
from dotenv import load_dotenv
import os
import requests
from bs4 import BeautifulSoup
from urlextract import URLExtract
from markdownify import markdownify as md

"""
This file contains all the functions needed to get the elements
from an manga used in the mangaSearch function in main.
"""

def getMangaUrl(keyword, position):
    """Takes in a keyword and a position, and returns the URL of the search result at that position

    Args:
        keyword (string): keyword to search
        position (int): position of which to return url of (top result, second result, etc.)

    Returns:
        string: URL of the search result at position "position"
    """

    # the index of the first search result in the array of links on the page is 93
    # there are 4 miscellaneous links between results, so for the next position, go 5 links down
    index = 92 + (5 * position)

    r = requests.get(f'https://myanimelist.net/manga.php?q={keyword}&cat=manga')
    soup = BeautifulSoup(r.text, 'html.parser')
    links = soup.find_all('a')

    return links[index].get('href')

def getMangaDescription(url):
    """Gets the description from a URL of a specific manga

    Args:
        url (string): URL to get description from

    Returns:
        string: the description field from URL input
    """
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    description_html = soup.find_all('span', attrs={'itemprop':'description'})

    #str(description_html)[27:-5].replace('<br>', '\n').replace('<br/>', '')
    description = str(description_html)[30:-8].replace('<br>', '').replace('</br>', '').replace('<br/>', '')

    return md(description)

def getMangaImage(url):
    """Returns the URL for the cover image of an manga

    Args:
        url (string): URL to take image from

    Returns:
        string: image URL to the .jpg cover image
    """
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    image_html = soup.find_all(attrs={'itemprop':'image'})

    extractor = URLExtract()
    image = extractor.find_urls(str(image_html[0]))

    return image[0]

ProfileSearch.py
import discord
from discord import SyncWebhook
from discord.utils import get
from dotenv import load_dotenv
import os
import requests
from bs4 import BeautifulSoup
from urlextract import URLExtract

"""
This file contains all the functions needed to get the elements
from a user profile used in the profileSearch function in main.
"""

def getDays(profile_url, index):
    """Returns an array of days of anime watched and days of manga read

    Args:
        profile_url (string): profile link to search
        index (int): whether to return anime or manga days

    Returns:
        list: list of the days, days[0] = anime days watched, days[1] = days of manga read
    """

    r = requests.get(profile_url)
    soup = BeautifulSoup(r.text, 'html.parser')
    beginning = '<div class="di-tc al pl8 fs12 fw-b"><span class="fn-grey2 fw-n">Days: </span>'
    end = '</div>'
    days = soup.find_all("div", {"class": "di-tc al pl8 fs12 fw-b"})

    days[0] = str(days[0]).replace(beginning, "").replace(end, "")
    days[1] = str(days[1]).replace(beginning, "").replace(end, "")

    return days[index]

def getEpisodes(profile_url):
    """Returns the number of episodes watched from a given profile

    Args:
        profile_url (string): profile link to search

    Returns:
        string: number of episodes watched
    """
    r = requests.get(profile_url)
    soup = BeautifulSoup(r.text, 'html.parser')
    beginning = '<span class="di-ib fl-r">'
    end = '</span>'
    episodes = soup.find_all("span", {"class": "di-ib fl-r"})
    episodes[2] = str(episodes[2]).replace(beginning, "").replace(end, "")

    return episodes[2]

def getWatching(profile_url):
    """Returns the number of anime currently watching

    Args:
        profile_url (string): profile link to search

    Returns:
        string: number of anime currently watching
    """
    r = requests.get(profile_url)
    soup = BeautifulSoup(r.text, 'html.parser')
    beginning = '<span class="di-ib fl-r lh10">'
    end = '</span>'
    watching = soup.find_all("span", {"class": "di-ib fl-r lh10"})
    watching[0] = str(watching[0]).replace(beginning, "").replace(end, "")

    return watching[0]

def getCompleted(profile_url):
    """Returns the number of completed anime

    Args:
        profile_url (_type_): _description_

    Returns:
        string: number of completed anime
    """
    r = requests.get(profile_url)
    soup = BeautifulSoup(r.text, 'html.parser')
    beginning = '<span class="di-ib fl-r lh10">'
    end = '</span>'
    watching = soup.find_all("span", {"class": "di-ib fl-r lh10"})
    watching[1] = str(watching[1]).replace(beginning, "").replace(end, "")

    return watching[1]

def getProfilePicture(profile_url):
    """Gets the profile picture of a given profile

    Args:
        profile_url (string): profile link to search

    Returns:
        string: url of the profile picture
    """
    r = requests.get(profile_url)
    soup = BeautifulSoup(r.text, 'html.parser')
    image_html = soup.find_all("div", {"class": "user-image"})

    extractor = URLExtract()
    image = extractor.find_urls(str(image_html[0]))

    return image[0]

AnimeSearch.py
import discord
from discord import SyncWebhook
from discord.utils import get
from dotenv import load_dotenv
import os
import requests
from bs4 import BeautifulSoup
from urlextract import URLExtract
from markdownify import markdownify as md

"""
This file contains all the functions needed to get the elements
from an anime used in the animeSearch function in main.
"""

def getUrl(keyword, position):
    """Takes in a keyword and a position, and returns the URL of the search result at that position

    Args:
        keyword (string): keyword to search
        position (int): position of which to return url of (top result, second result, etc.)

    Returns:
        string: URL of the search result at position "position"
    """

    # the index of the first search result in the array of links on the page is 93
    # there are 4 miscellaneous links between results, so for the next position, go 5 links down
    index = 92 + (5 * position)

    r = requests.get(f'https://myanimelist.net/anime.php?q={keyword}&cat=anime')
    soup = BeautifulSoup(r.text, 'html.parser')
    links = soup.find_all('a')

    return links[index].get('href')

def getDescription(url):
    """Gets the description from a URL of a specific anime

    Args:
        url (string): URL to get description from

    Returns:
        string: the description field from URL input
    """
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    description_html = soup.find_all('p', attrs={'itemprop':'description'})

    description = str(description_html)[27:-5].replace('<br>', '').replace('</br>', '').replace('<br/>', '')

    return md(description)

def getImage(url):
    """Returns the URL for the cover image of an anime

    Args:
        url (string): URL to take image from

    Returns:
        string: image URL to the .jpg cover image
    """
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    image_html = soup.find_all(attrs={'itemprop':'image'})

    extractor = URLExtract()
    image = extractor.find_urls(str(image_html[0]))

    return image[0]